### ISSUE TITLE
Crash from lazygit with $GIT_DIR set and $GIT_WORK_TREE unset

### ISSUE THREAD

**Issue by jwhitley** [jesseduffield/lazygit#3175]
**Describe the bug**
I'm seeing a crash from lazygit under a working scenario similar to ones described in #1294.  Specifically, lazygit appears to crash if $GIT_DIR is set, but $GIT_WORK_TREE (or --work-tree) is not. This scenario is supported by git, as I'll describe below.  In my specific case, the bare repo actually does set `core.worktree` but lazygit is unaware of that method of setting the work tree.

**To Reproduce**

1. Create a bare repo: `git init --bare <some_tmp_repo>`
2. In another location, create a "work tree" directory and `cd` into it.
3. Set `$GIT_DIR` to the path of the repo created in step 1
4. Invoke `lazygit`

**Expected behavior**
Lazygit starts, showing the expected git status of the bare repo versus the work tree directory in step 2.

**Actual behavior**
lazygit (0.40.2) crashes with output:

```
2023/12/20 16:53:56 An error occurred! Please create an issue at: https://github.com/jesseduffield/lazygit/issues

*errors.errorString Error getting repo paths: failed to get worktree git dir path: stat /Users/whitley/.git: no such file or directory
github.com/jesseduffield/lazygit/pkg/commands/git.go:100 (0x1029832b8)
github.com/jesseduffield/lazygit/pkg/gui/gui.go:271 (0x102aa92ac)
github.com/jesseduffield/lazygit/pkg/gui/gui.go:656 (0x102aabc00)
github.com/jesseduffield/lazygit/pkg/gui/gui.go:675 (0x102aac200)
github.com/jesseduffield/lazygit/pkg/utils/utils.go:108 (0x102826960)
github.com/jesseduffield/lazygit/pkg/gui/gui.go:674 (0x102aac16c)
github.com/jesseduffield/lazygit/pkg/app/app.go:263 (0x102ad44bc)
github.com/jesseduffield/lazygit/pkg/app/app.go:48 (0x102ad4481)
github.com/jesseduffield/lazygit/pkg/app/entry_point.go:150 (0x102ad608c)
github.com/jesseduffield/lazygit/main.go:23 (0x102ad7858)
runtime/internal/atomic/types.go:194 (0x1024b2afc)
runtime/asm_arm64.s:1197 (0x1024e4434)
```

**Version info:**
lazygit version: commit=, build date=, build source=homebrew, version=0.40.2, os=darwin, arch=arm64, git version=2.43.0
git version 2.43.0

**Additional context**
From the [git config docs for `core.worktree`](https://git-scm.com/docs/git-config#Documentation/git-config.txt-coreworktree):

> [...] The value can be an absolute path or relative to the path to the .git directory, which is either specified by --git-dir or GIT_DIR, or automatically discovered. If --git-dir or GIT_DIR is specified but none of --work-tree, GIT_WORK_TREE and core.worktree is specified, the current working directory is regarded as the top level of your working tree.

The key part are the fallbacks: git looks for --work-tree, GIT_WORK_TREE, and core.worktree, defaulting to `cwd` if none of those are available.  lazygit either needs to implement work tree identification as git does, or simply delegate that to git itself by calling `git rev-parse --show-toplevel`. That is the [canonical CLI call in the git codebase to retrieve git's idea of the work tree](https://github.com/git/git/blob/055bb6e9969085777b7fab83e3fee0017654f134/builtin/rev-parse.c#L927).

Fixing this issue as I describe above will also fix #1294 in passing, since pretty much all of those problems are tied to these problems with work tree identification.

**User scenario**
I'm using [vcsh](https://github.com/RichiH/vcsh) to version my homedir files.  vcsh is a tool that manages overlaying several bare git repos, so that different concerns (e.g. shell, nvim, platform-specific configs, etc.) can be versioned separately. It's used from the command line as `vcsh <repo_name>` to enter a subshell with $GIT_DIR set (to `.config/vcsh/repo.d/<repo_name>.git`). vcsh sets the git config `core.worktree` in each bare repo. This is a standard approach to specifying a bare tree's work tree in this and similar scenarios.

**Comment by jwhitley** [comment]
Follow up: I have an [early WIP branch on this issue](https://github.com/jwhitley/lazygit/tree/jwhitley/work-tree-from-git). It's not ready for PR, but the proof-of-concept code works to some extent. (That branch was added/committed/pushed with itself!)  Next step is to get the tests sorted out. Since this is a significant change to the affected code's dependencies, that's a bit of work to adapt the tests re: the switch from afero to OsCommand.

**Comment by jwhitley** [comment]
Fixed in #3183.
